File: benchpress/benchmarks/gameoflife/python_numpy/gameoflife.py
from __future__ import print_function

"""
Game of Life
------------

"""
from benchpress.benchmarks import util
import numpy as np

bench = util.Benchmark("Game of Life", "height*width*iterations*rules")

SURVIVE_LOW = 2
SURVIVE_HIGH = 3
SPAWN = 3


def world(height, width):
    state = np.ones((height + 2, width + 2), dtype=bench.dtype)
    state[2:-2, 2:-2] = bench.dtype(0)
    return state


def world_zeros(height, width):
    state = np.zeros((height + 2, width + 2), dtype=bench.dtype)
    return state


def play(state, iterations, version=1, visualize=False):
    cells = state[1:-1, 1:-1]
    ul = state[0:-2, 0:-2]
    um = state[0:-2, 1:-1]
    ur = state[0:-2, 2:]
    ml = state[1:-1, 0:-2]
    mr = state[1:-1, 2:]
    ll = state[2:, 0:-2]
    lm = state[2:, 1:-1]
    lr = state[2:, 2:]

    def update():
        """
        This is the first implementation of the game rules.
        """
        neighbors = ul + um + ur + ml + mr + ll + lm + lr  # count neighbors
        live = neighbors * cells  # extract live cells neighbors
        stay = (live >= SURVIVE_LOW) & (live <= SURVIVE_HIGH)  # find cells the stay alive
        dead = neighbors * (cells == 0)  # extract dead cell neighbors
        spawn = dead == SPAWN  # find cells that spaw new life

        cells[:] = stay | spawn  # save result for next iteration

    def update_optimized():
        """
        This is an optimized implementation of the game rules.
        """
        neighbors = ul + um + ur + ml + mr + ll + lm + lr  # Count neighbors

        c1 = (neighbors == SURVIVE_LOW)  # Life conditions
        c2 = (neighbors == SPAWN)

        cells[:] = cells * c1 + c2  # Update

    if version == 1:  # Select the update function
        update_func = update
    elif version == 2:
        update_func = update_optimized

    for i in range(iterations):  # Run the game
        if visualize:
            bench.plot_surface(state, "3d", 16, 1, 0)
        update_func()
        bench.flush()

    return state


def main():
    (H, W, I, V) = bench.args.size

    if V not in [1, 2]:
        raise Exception("Unsupported rule-implementation.")

    S = world(H, W)

    bench.start()
    R = play(S, I, V, bench.args.visualize)
    bench.stop()
    bench.save_data({'res': R})
    bench.pprint()


if __name__ == "__main__":
    main()

File: benchpress/benchmarks/gameoflife/cpp11_bxx/src/gameoflife.cpp
#include <iostream>
#include <iomanip>
#include <bxx/bohrium.hpp>
#include <bp_util.h>

static double SURVIVE_LOW  = 2.0;
static double SURVIVE_HIGH = 3.0;
static double SPAWN        = 3.0;

using namespace std;
using namespace bxx;

template <typename T>
void world(int height, int width, multi_array<T>& state)
{
    state = ones<T>(height+2, width+2);
    state[_(2,-3)][_(2,-3)] = (T)0;
}

template <typename T>
void play(multi_array<T>& state, int iterations, int version, int visualize)
{
    multi_array<T> cells, ul, um, ur, ml, mr, ll, lm, lr;
    cells   = state[_(1,-2)][_(1,-2)];
    ul      = state[_(0,-3)][_(0,-3)];
    um      = state[_(0,-3)][_(1,-2)];
    ur      = state[_(0,-3)][_(2,-1)];
    ml      = state[_(1,-2)][_(0,-3)];
    mr      = state[_(1,-2)][_(2,-1)];
    ll      = state[_(2,-1)][_(0,-3)];
    lm      = state[_(2,-1)][_(1,-2)];
    lr      = state[_(2,-1)][_(2,-1)];

    multi_array<T> neighbors;
    if (1 == version) {             //  This is the first implementation of the game rules.
        multi_array<T> live, dead;
        multi_array<bool> stay, spawn;
        for (int i=0; i<iterations; ++i) {
            neighbors   = ul + um + ur + ml + mr + ll + lm + lr;// Count neighbors
            live        = neighbors * cells;                    // Extract live cells neighbors
                                                                // Find cells that stay alive
            stay        = (SURVIVE_LOW <= live) && 
                          (live <= SURVIVE_HIGH);  
            dead        = neighbors * as<T>(cells == (T)0);     // Extract dead cell neighbors
            spawn       = dead == SPAWN;                        // Find spawning cells
            
            cells(as<T>(stay || spawn));                        // Update state

            if (visualize) {
                plot_surface(state, 1, 16, 1, 0);
            }
        }
    } else if (2 == version) {      // This is an optimized version of the game rules
        multi_array<T> c1, c2;
        for (int i=0; i<iterations; ++i) {
            neighbors = ul + um + ur + ml + mr + ll + lm + lr;  // Count neighbors

            c1 = as<T>(neighbors == SURVIVE_LOW);               // Life conditions
            c2 = as<T>(neighbors == SPAWN);

            cells(cells * c1 + c2);                             // Update

            if (visualize) {
                plot_surface(state, 1, 16, 1, 0);
            }
        }
    } else {
        throw std::runtime_error("Unsupported version.");
    }
}

template <typename T>
void bench(bp_util_type& bp, const int W, const int H, const int I, const int V)
{
    multi_array<T> state;                           // Construct matrices
    world(H, W, state);

    Runtime::instance().flush();
    bp.timer_start();                               // Start timer

    play(state, I, V, bp.args.visualize);

    Runtime::instance().flush();
    bp.timer_stop();                                // Stop timer

    bp.print("gameoflife(cpp11_bxx)");				// Print results..
    if (bp.args.verbose) {                          // ..and value.
        cout << fixed << setprecision(10)
             << "Result = " << scalar(sum(state)) << endl;
    }
}

int main(int argc, char* argv[])
{
    bp_util_type bp = bp_util_create(argc, argv, 4);// Grab arguments
    if (bp.args.has_error) {
        return 1;
    }
    const int W = bp.args.sizes[0];
    const int H = bp.args.sizes[1];
    const int I = bp.args.sizes[2];
    const int V = bp.args.sizes[3];

    bench<double>(bp, W, H, I, V);

    return 0;
}
